# Task 1 – NumPy & Pandas Basics

## 1. NumPy Basics

In [16]:
import numpy as np

# Array creation
arr = np.array([10, 20, 30, 40, 50])
print("Array:", arr)

# Indexing
print("First element:", arr[0])
print("Third element:", arr[2])

# Slicing
print("First three elements:", arr[:3])
print("Last two elements:", arr[-2:])

# Broadcasting
arr2 = arr + 10
print("After broadcasting:", arr2)

Array: [10 20 30 40 50]
First element: 10
Third element: 30
First three elements: [10 20 30]
Last two elements: [40 50]
After broadcasting: [20 30 40 50 60]


In [17]:
# 2D Array
matrix = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
])

print("Matrix:")
print(matrix)

print("First row:", matrix[0])
print("Second column:", matrix[:, 1])

Matrix:
[[1 2 3]
 [4 5 6]
 [7 8 9]]
First row: [1 2 3]
Second column: [2 5 8]


## 2. Pandas DataFrame Analysis

In [18]:
import pandas as pd

url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"

df = pd.read_csv(url)

df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [19]:
print("Shape:", df.shape)
print("Columns:")
print(df.columns)

df.info()

Shape: (891, 12)
Columns:
Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


## 3. Indexing with loc[] and iloc[]

In [20]:
# loc - label based selection
print(df.loc[0, "Name"])

# iloc - position based selection
print(df.iloc[0, 3])

# Select multiple columns
df.loc[:, ["Name", "Age", "Sex", "Survived"]].head()

Braund, Mr. Owen Harris
Braund, Mr. Owen Harris


,Name,Age,Sex,Survived
0,"Braund, Mr. Owen Harris",22.0,male,0
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",38.0,female,1
2,"Heikkinen, Miss. Laina",26.0,female,1
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0,female,1
4,"Allen, Mr. William Henry",35.0,male,0


## 4. Filtering

In [21]:
# Passengers who survived
survived = df[df["Survived"] == 1]

print("Survived passengers:", len(survived))

Survived passengers: 342


In [22]:
# Female passengers
female_passengers = df[df["Sex"] == "female"]

print("Female passengers:", len(female_passengers))

Female passengers: 314


In [23]:
# Passengers older than 30
older_passengers = df[df["Age"] > 30]

older_passengers.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
11,12,1,1,"Bonnell, Miss. Elizabeth",female,58.0,0,0,113783,26.5500,C103,S


## 5. GroupBy Analysis

In [24]:
grouped = df.groupby("Sex")["Survived"].mean()

print(grouped)

Sex
female    0.742038
male      0.188908
Name: Survived, dtype: float64


In [25]:
class_analysis = df.groupby("Pclass").agg(
    Total_Passengers=("PassengerId", "count"),
    Average_Age=("Age", "mean"),
    Survival_Rate=("Survived", "mean")
)

class_analysis

,Total_Passengers,Average_Age,Survival_Rate
Pclass,,,
1,216,38.233441,0.629630
2,184,29.877630,0.472826
3,491,25.140620,0.242363


## 6. Aggregations

In [26]:
aggregations = df["Age"].agg([
    "count",
    "mean",
    "median",
    "min",
    "max"
])

aggregations

,Age
count,714.000000
mean,29.699118
median,28.000000
min,0.420000
max,80.000000


In [27]:
class_info = pd.DataFrame({
    "Pclass": [1, 2, 3],
    "Class_Type": ["First Class", "Second Class", "Third Class"]
})

class_info

,Pclass,Class_Type
0,1,First Class
1,2,Second Class
2,3,Third Class


In [28]:
merged_df = pd.merge(
    df,
    class_info,
    on="Pclass",
    how="left"
)

merged_df[["Name", "Pclass", "Class_Type"]].head()

,Name,Pclass,Class_Type
0,"Braund, Mr. Owen Harris",3,Third Class
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1,First Class
2,"Heikkinen, Miss. Laina",3,Third Class
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1,First Class
4,"Allen, Mr. William Henry",3,Third Class


## 7. Key Findings

The Titanic dataset contains 891 passenger records.
Female passengers had a much higher survival rate than male passengers.
First-class passengers had a higher survival rate than second- and third-class passengers.
Passenger age varied across the dataset, with some missing age values.
GroupBy analysis helped compare survival across gender and passenger class.
Pandas merge was used to add descriptive information about passenger classes.